src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+    }
+
+    private Triple<String, String, String> cacheKey(String ns, OwnerReference resource) {
+        return Triple.of(ns, resource.getKind(), resource.getName());
+    }
+
+    // Unfortunately, ObjectReference and OwnerReference both independently implement getKind and
+    // getName - they don't come from a common base class.
+    private Triple<String, String, String> cacheKey(String ns, ObjectReference resource) {
+        return Triple.of(ns, resource.getKind(), resource.getName());
+    }
+
+    private Pair<HasMetadata, DiscoveryNode> queryForNode(
+            Triple<String, String, String> lookupKey) {
+
+        String namespace = lookupKey.getLeft();
+        String name = lookupKey.getRight();
+        KubeDiscoveryNodeType nodeType =
+                KubeDiscoveryNodeType.fromKubernetesKind(lookupKey.getMiddle());
+        if (nodeType == null) {
+            return null;
+        }
+
+        synchronized (queryLocks.computeIfAbsent(lookupKey, k -> new Object())) {
+            HasMetadata kubeObj =
+                    nodeType.getQueryFunction().apply(client()).apply(namespace).apply(name);
+
+            DiscoveryNode node =
+                    DiscoveryNode.getNode(
+                                    n -> {
+                                        return name.equals(n.name)
+                                                && namespace.equals(
+                                                        n.labels.get(
+                                                                DISCOVERY_NAMESPACE_LABEL_KEY));
+                                    })
+                            .orElseGet(
+                                    () -> {
+                                        DiscoveryNode newNode = new DiscoveryNode();
+                                        newNode.name = name;
+                                        newNode.nodeType = nodeType.getKind();
+                                        newNode.children = new ArrayList<>();
+                                        newNode.target = null;
+                                        newNode.labels =
+                                                kubeObj != null
+                                                        ? kubeObj.getMetadata().getLabels()
+                                                        : new HashMap<>();
+                                        // Add namespace to label to retrieve node later
+                                        newNode.labels.put(
+                                                DISCOVERY_NAMESPACE_LABEL_KEY, namespace);
+                                        return newNode;
+                                    });
+            return Pair.of(kubeObj, node);
+        }
+    }
+
+    @ApplicationScoped
+    static final class KubeConfig {
+        public static final String KUBERNETES_NAMESPACE_PATH =
+                "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
```

Comment:
I think this could also be a `@ConfigProperty`, but this path should be used as the default value in `application.properties`.